### PR TITLE
Further refactor `ImportanceEvaluator`s

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
This is a follow-up PR to #3552. By slightly breaking backward compatibility of the interface of `ImportanceEvaluator`s, we could make further refactoring.

## Description of the changes
We change the interface of `BaseImportanceEvaluator` such that the arguments `params` and `target` do not get `None` values. We fill the default `params` and `target` in `get_param_importances()` function, rather than in each `ImportanceEvaluator`. This makes it easier to implement new `ImportanceEvaluator`, as well as keeps the default value code in one place.

We also gather the code for sorting results in `get_param_importances()`, such that `ImportanceEvaluator`s do not need to return a sorted `OrderedDict`.

This breaks the interface of `ImportanceEvaluator.evaluate()` function if they are directly called from user code. However, the documentation explicitly states that users should not call that function directly, thus I don't think that is a problem.

This also breaks the type integrity of user-defined `ImportanceEvaluator`s. However, there won't be any runtime errors as long as they are called from `get_param_importances()`, since the domain of `ImportanceEvaluator.evaluate` will become smaller.

### CODE REVIEW

**Comment by c-bata** [comment]
Should we make this draft until #3552 is merged?

**Comment by contramundum53** [comment]
This PR is review-ready now. Could you take a look? @c-bata @nzw0301

File: optuna/importance/__init__.py [hunk 1]
```diff
@@ -78,13 +81,45 @@ def get_param_importances(
 
     Returns:
         An :class:`collections.OrderedDict` where the keys are parameter names and the values are
-        assessed importances.
+        assessed importances. The importances will be sorted in a descending order.
 
     """
+
+    if len(study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,))) == 0:
+        raise ValueError("The study does not contain completed trials.")
```
Comment (c-bata) [inline optuna/importance/__init__.py:89]: I guess this check actually causes performance degradation of optuna-fast-fanova. Can we avoid to issue an additional SQL query here?

Comment (contramundum53) [reply optuna/importance/__init__.py:89]: In my understanding, if we use `CachedStorage`, the results of `get_trials` are cached so that we won't need to fetch all data every time.

Comment (c-bata) [reply optuna/importance/__init__.py:89]: Yes, it's true if we use CachedStorage.

Comment (contramundum53) [reply optuna/importance/__init__.py:89]: We could change the first argument of `BaseImportanceEvaluator.evaluate` from `study` to `trials: List[FrozenTrial]`, and this would remove the duplicate of both SQL commands and the code. However, I think it is good to do that in a followup PR, since that is going to be another large change. What do you think? @c-bata

**Comment by c-bata** [comment]
@toshihikoyanase Could you review this PR?

File: optuna/importance/__init__.py (hunk 1 above)
Comment (contramundum53) [reply optuna/importance/__init__.py:89]: #3664 Here is the followup PR :)

File: tests/integration_tests/test_shap.py [hunk 2]
```diff
@@ -194,7 +211,7 @@ def objective(trial: Trial) -> Tuple[float, float]:
         study.optimize(objective, n_trials=3)
 
         with pytest.raises(ValueError):
-            ShapleyImportanceEvaluator().evaluate(study)
+            get_param_importances(study, evaluator=ShapleyImportanceEvaluator())
```
Comment (c-bata) [inline tests/integration_tests/test_shap.py:214]: I'd say it's better to focus the behavior of `ShapleyImportanceEvaluator().evaluate(study)` in this test file. So how about calling `ShapleyImportanceEvaluator().evaluate(study)` directly?

File: optuna/importance/__init__.py [hunk 3]
```diff
@@ -68,8 +71,8 @@ def get_param_importances(
         target:
             A function to specify the value to evaluate importances.
             If it is :obj:`None` and ``study`` is being used for single-objective optimization,
-            the objective values are used. ``target`` must be specified if ``study`` is being
-            used for multi-objective optimization.
```
[the remark this answers is not in the document]
Comment (contramundum53) [reply optuna/importance/__init__.py:72]: It is already written in the note right below, so I thought it would be a duplicate and removed it.

File: tests/integration_tests/test_shap.py (hunk 2 above)
Comment (contramundum53) [reply tests/integration_tests/test_shap.py:214]: `ShapleyImportanceEvaluator().evaluate` does not handle errors anymore, so we cannot directly test it for invalid cases.
Maybe we do not need these tests because the error-handling behavior is already tested in `tests/importance_tests/test_init.py`.

File: tests/integration_tests/test_botorch.py [hunk 4]
```diff
@@ -78,9 +79,9 @@ def candidates_func(
         train_obj: torch.Tensor,
         train_con: Optional[torch.Tensor],
         bounds: torch.Tensor,
-    ) -> torch.Tensor:
+    ) -> Any:
         # Must be a `torch.Tensor`, not a list.
-        return torch.rand(1).tolist()
+        return [torch.rand(1).tolist()]
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_botorch.py:84]: I guess this change is not related to refactoring of `ImportanceEvaluator`. Could you revert it, please?

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
I'm trying to catch up with the series of the changes. Let me share my early comment.

File: tests/integration_tests/test_botorch.py (hunk 4 above)
Comment (contramundum53) [reply tests/integration_tests/test_botorch.py:84]: Without this change, mypy fails.

Comment (toshihikoyanase) [reply tests/integration_tests/test_botorch.py:84]: We'll work on this change in https://github.com/optuna/optuna/pull/3679. So, could you revert the change in this PR?

File: optuna/importance/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/importance/__init__.py:72]: Thanks! You are correct. I missed that :bow:

File: tests/integration_tests/test_shap.py (hunk 2 above)
Comment (c-bata) [reply tests/integration_tests/test_shap.py:214]: Agree 👍  I think we can simply remove these tests.

File: tests/integration_tests/test_botorch.py (hunk 4 above)
Comment (contramundum53) [reply tests/integration_tests/test_botorch.py]: I reverted this change. PTAL.

**Comment by nzw0301** [comment]
@contramundum53  Could you resolve conflicts?

**Comment by contramundum53** [comment]
Fixed.

**Review by c-bata — APPROVED** [review]
Thanks for the update! LGTM.

File: tests/integration_tests/test_botorch.py (hunk 4 above)
Comment (toshihikoyanase) [reply tests/integration_tests/test_botorch.py]: Thanks.

**Comment by toshihikoyanase** [comment]
I'm sorry for the delayed response. Could you resolve the conflict, please?

**Comment by contramundum53** [comment]
Thanks, fixed.

**Comment by toshihikoyanase** [comment]
The importance tests failed. Could you check them?
```
=========================== short test summary info ============================
FAILED tests/importance_tests/test_init.py::test_importance_evaluator_seed[MeanDecreaseImpurityImportanceEvaluator]
FAILED tests/importance_tests/test_init.py::test_importance_evaluator_seed[FanovaImportanceEvaluator]
FAILED tests/importance_tests/test_init.py::test_importance_evaluator_seed[ShapleyImportanceEvaluator]
FAILED tests/importance_tests/test_init.py::test_importance_evaluator_with_target[MeanDecreaseImpurityImportanceEvaluator]
FAILED tests/importance_tests/test_init.py::test_importance_evaluator_with_target[FanovaImportanceEvaluator]
FAILED tests/importance_tests/test_init.py::test_importance_evaluator_with_target[ShapleyImportanceEvaluator]
```

https://github.com/optuna/optuna/runs/7492253049?check_suite_focus=true
